cmd/commands.go

```diff
@@ -0,0 +1,193 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/StackExchange/dnscontrol/models"
+	"github.com/urfave/cli"
+)
+
+// categories of commands
+const (
+	catMain  = "main"
+	catDebug = "debug"
+	catUtils = "utility"
+)
+
+// Run will execute the CLI
+func Run(version string) error {
+	app := cli.NewApp()
+	app.Version = version
+	app.Name = "dnscontrol"
+	app.HideVersion = true
+	app.Usage = "dnscontrol is a compiler and dsl for managing cloud dns zones"
+	app.Commands = []cli.Command{
+		*previewCommand,
+		*pushCommand,
+		*debugJSCommand,
+		*debugPreprocessCommand,
+		cli.Command{
+			Name:  "version",
+			Usage: "Print version information",
+			Action: func(c *cli.Context) {
+				fmt.Println(version)
+			},
+			Category: catDebug,
+		},
+	}
+	app.EnableBashCompletion = true
+	app.Run(os.Args)
+	return nil
+}
+
+// Shared config types
+
+// GetDNSConfigArgs contains what we need to get a valid dns config.
+// Could come from parsing js, or from stored json
+type GetDNSConfigArgs struct {
+	ExecuteDSLArgs
+	JSONFile string
+}
+
+func (args *GetDNSConfigArgs) flags() []cli.Flag {
+	return append(args.ExecuteDSLArgs.flags(),
+		cli.StringFlag{
+			Destination: &args.JSONFile,
+			Name:        "json",
+			Usage:       "file containing intermediate json",
+		},
+	)
+}
+
+func GetDNSConfig(args GetDNSConfigArgs) (*models.DNSConfig, error) {
+	if args.JSONFile != "" {
+		f, err := os.Open(args.JSONFile)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+		dec := json.NewDecoder(f)
+		cfg := &models.DNSConfig{}
+		if err = dec.Decode(cfg); err != nil {
+			return nil, err
+		}
+		return cfg, nil
+	}
+	return ExecuteDSL(args.ExecuteDSLArgs)
+}
+
+// ExecuteDSLArgs are used anytime we need to read and execute dnscontrol javascript
+type ExecuteDSLArgs struct {
+	JSFile   string
+	JSONFile string
+	DevMode  bool
+}
+
+func (args *ExecuteDSLArgs) flags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:        "js",
+			Value:       "dnsconfig.js",
+			Destination: &args.JSFile,
+			Usage:       "Javascript file containing dns config",
```

Comment:
Main config file: Describes DNS zones, records, providers and other info.

Can this flag be "-js" for backwards compatibility but also something more descriptive like "-config"?